lib/khovratovich/index.js

```diff
@@ -1,40 +1,162 @@
+  if((0xffffffff - currentNonce) < (maxNonces - 1)) {
+    return callback(
+      new Error('Equihash \'nonce\' limited to 32 bit values.'));
+  }
+
+  addon.solve({n, k, personal, seed, nonce, maxNonces}, function(err, proof) {
+    //console.log(err, proof, proof.value.length);
+    if(err) {
+      return callback(err);
+    }
+
+    callback(null, {
+      n: proof.n,
+      k: proof.k,
+      personal: proof.personal,
+      nonce: proof.nonce,
+      solution: proof.solution});
+  });
 };
 
-exports.verify = (input, options) => {
-  const parameters = {
-    n: options.n || 90,
-    k: options.k || 5,
-    nonce: options.nonce || 1,
-    seed: input,
-    value: options.value
-  };
+exports.verify = (seed, options, callback) => {
+  const n = 'n' in options ? options.n : 90;
+  const k = 'k' in options ? options.k : 5;
+  const nonceLength = 'nonceLength' in options ? options.nonceLength : 4;
+  const personal = 'personal' in options ? options.personal : Buffer.alloc(exports.PERSONALBYTES, 0);
+  const solution = options.solution;
+
+  // k must be in range of [1,7]
+  if(!(k >= 1 && k <= 7)) {
+    return callback(
+      new Error('Equihash \'k\' parameter must be from 1 to 7.'));
+  }
+
+  if((n / (k + 1)) > 32) {
+    return callback(
+      new Error('Equihash \'n\' and \'k\' must satisfy n/(k+1) <= 32.'));
+  }
+
+  if(personal.length > exports.PERSONALBYTES) {
+    return callback(
+      new Error('Equihash \'personal\' length too large.'));
+  }
+
+  // convert nonce to buffer
+  let nonce = options.nonce;
+  if(typeof nonce === 'number') {
+    const ab = new ArrayBuffer(nonceLength);
+    const b = Buffer.from(ab);
+    b.fill(0);
+    const dv = new DataView(ab);
+    dv.setUint32(0, nonce, true);
+    nonce = b;
+  } else if(!(nonce instanceof Buffer)) {
+    return callback(
+      new Error('Equihash \'nonce\' must be buffer or number.'));
+  }
+
+  // solution is array of 2^k 32 bit unsigned int values
+  // each array element only uses n/(k+1) + 1 bits
+  if(solution.length !== Math.pow(2,k)) {
+    return callback(
+      new Error('Equihash \'solution\' is not 2^k 32 bit values.'));
+  }
 
-  if(parameters.value.length < 128) {
-    // solutions less than 128 bytes in length are invalid
-    return false;
+  // distinct indices check
+  // missing from khovratovich implementation
+  if(solution.length !== new Set(solution).size) {
+    return callback(null, false);
+    //return callback(new Error('Equihash indices not distinct.'));
   }
 
-  if(parameters.k < 1 || parameters.k > 7) {
-    // k must be less than 7
-    return false;
+  // ordered indices check
+  // missing from khovratovich implementation
+  // array is a tree structure, check left nodes < right nodes
+  for(let _k = 0; _k < k; ++_k) {
+    const stride = 1 << _k;
+    for(let i = 0; i < solution.length; i += (2 * stride)) {
+      if(solution[i] >= solution[i + stride]) {
+        return callback(null, false);
```

Comment:
Why false here instead of an invalid solution? Where is the line between an "invalid" solution and a "false" one? I would have thought that a solution that is properly ordered, has no duplicates -- but is just wrong, is "false" ... anything else is invalid and generates an error. Not saying we have to do it that way, just saying that was my expectation.

Reply:
It's open for debate where the line should be between an error and just a false return value.  Or if it should all be one way or the other.  Currently errors are mostly out of range or similar vs just bad solution data.  Can revisit this as needed.